### PR TITLE
[WIP] Action based menu handling

### PR DESCRIPTION
**Pull request by dkwingsmt**
This is a refactor that manipulates menus via intents and actions. The goal is to incorporate with the design for animated menus that uses dual menu controllers and no decorators or delegates, as described in [this comment](https://github.com/flutter/flutter/pull/165556#issuecomment-2792268796).

Key points:
* First, let's only look at `RawMenuAnchor`s:
  * The `MenuController` no longer directly calls `_RawMenuAnchorBaseMixin`'s methods. Instead, it dispatches `OpenMenuIntent`, `CloseMenuIntent`, etc.
    * The intents will be dispatched to a context that is assigned when the controller is attached, which is the context within the `Actions` widget of the raw menu anchor.
  * The raw menu anchor builds an `Actions` widget, which handles the intents with callback actions that call private methods, such as `open()` and `close()`.
  * This solves the problem of how the `MenuController`'s methods affect the raw menu anchor.
* As for themed menus:
  * As said above, the themed menu is given an external menu controller, which builds another (internal) menu controller for its raw menu anchor.
  * The themed menu will build its own `Actions` widget and handles opening and closing.
    * Opening will making the internal menu controller open and play animations, while closing will play animations before making the internal menu controller close.
    * Dismissing will be discussed later.
    * Similarly, the context just within _this_ `Actions` widget is assigned to the external menu controller. Therefore when calling the methods of the _external_ menu controller will eventually call private methods of the themed menu.
* How dismissing works:
  * The goal is that `DismissIntent`, the standard way of dismissing, such as pressing Esc, must be handled by the root themed menu.
  * This consists of two steps:
    * The `DismissIntent` is handled by any raw menu anchor, which finds its root raw menu anchor and call its `_dismiss()`.
    * The `_dismiss()` actually dispatches a `DismissMenuIntent`, which is by default handled by the raw menu anchor in an overridable way. The themed menu also handles it in its `Actions` by calling the animated `close()`.
    * We can't let `DismissIntent` be handled overridable because it would leak to the next widget surrounding it (such as a dialog barrier).

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by davidhicks980** [comment]
I'll post the same comment on this PR and the delegate PR so that viewers of the other PR have some context.

I spent some time playing with this PR last night. It's fundamentally similar to the decorator, but I liked your idea of themed menus overriding the actions of their RawMenuAnchor, because it feels like it fits with Flutter paradigms. I ended up with something completely different, but in the spirit of this PR. Here's my thought process:

What I like about the PR:
- It's more understandable than the decorator.
- It avoids the animationStatus errors that I introduced
- I like the concept of placing an intent at a different layer of the tree. I'm used to moving actions around, so this was a cool solution.
- Intent/Actions made the code a lot more explicit

A few issues that I came across:
- Passing BuildContext is usually discouraged because a BuildContext can be changed if a widget is reparented, and there's a risk of memory leaks. Also, exposing the attachment/detachment process increases the risk of memory leaks
- It's a lot of code to reimplement for each animated menu, and it's somewhat easy to mess up (e.g. what if you forget an action handler)
- Having MenuController.dismiss(), MenuController.close(), and MenuController.detach() would probably be confusing to someone who is first using MenuController.
- Existing menus (e.g. PopupMenuButton, context menus, etc) should probably be wired up to handle the new intents, otherwise users may be confused.
- Intent bubbling
  - For example, given a root menu and a submenu, if the internal submenu emits a DismissMenuIntent that is not overridden by the themed submenu, but is overridden by the root menu, wouldn't the root menu animate closed instead of the submenu synchronously closing? I'm not sure, but it's a bit confusing for me to think about.
- Less important because I think we were going to remove the animationStatus, but making the animation status mutable by everyone could lead to bugs.

So, I tried a couple different changes to simplify things. For example, we could pass a "inner" MenuController inside of the OpenMenuIntent/CloseMenuIntent that could be used by actions to perform a different opening or closing action. This would remove the need for attachment/detachment and BuildContext references. 

However, after thinking about it more, I realized our solutions are basically trying to accomplish the following:

1. We want MenuController.open() and MenuController.close() to animate by default
2. We want to provide a way to show or hide the overlay before and after animations using the MenuController
3. Allowing end users to skip menu animations could be useful (e.g. if you wanted to hide the menu immediately when transitioning to a new page)

I'm curious of your thoughts on this but what if skipped actions and decorators, and just added a flag to MenuController.open() and MenuController.close() that chose whether to trigger animation callbacks on RawMenuAnchor. For example, if `transition` was the flag and it was true by default, MenuController.open() would trigger onOpenRequested and MenuController.open(transition: false) would hide the overlay. The PR is here: https://github.com/flutter/flutter/pull/167806.

```dart
final MenuController menuController = MenuController;
late final AnimationController animationController = AnimationController(vsync: this, duration: Duration(milliseconds: 200));

void _handleCloseRequest() {
    animationController.reverse().whenComplete(() {
      menuController.close(transition: false); // Hides the overlay without calling onCloseRequest
    });
  }

  void _handleOpenRequest() {
    menuController.open(transition: false); // Shows the overlay without calling onOpenRequested
    animationController.forward();
  }

  // Example function called by the anchor
  void _handleAnchorTapped() {
     if (animationController.isForwardOrCompleted) {
        menuController.close();
      } else {
        menuController.open();
      }
   }

  @override
  Widget build(BuildContext context) {
    return RawMenuAnchor(
      controller: menuController,
      onCloseRequested: _handleCloseRequest,  // Called when MenuController.close() is called
      onOpenRequested: _handleOpenRequest.   // Called when MenuController.open() is called
     ...
  )
)
```

Let me know! Sorry, I realize this is has been a drawn out discussion.

**Comment by dkwingsmt** [comment]
Closing this PR in favor of https://github.com/flutter/flutter/pull/167806.
